ggml-cuda.cu

```diff
@@ -4197,9 +4197,53 @@ static void quantize_row_q8_1_cuda(const float * x, void * vy, const int kx, con
     quantize_q8_1<<<num_blocks, block_size, 0, stream>>>(x, vy, kx, kx_padded);
 }
 
+static __device__ __forceinline__ dfloat2 dfmul2(dfloat2 a, dfloat2 b) {
+#ifdef GGML_CUDA_F16
+    return __hmul2(a, b);
+#else
+    return make_float2(a.x * b.x, a.y * b.y);
+#endif
+}
+
+static __device__ __forceinline__ float2 dfloat22float2(dfloat2 a) {
+#ifdef GGML_CUDA_F16
+    return __half22float2(a);
+#else
+    return a;
+#endif
+}
+
+static __global__ void dequantize_block_q4_0(const void * __restrict__ vx, float * __restrict__ y, const int k) {
+    const int i = blockDim.x*blockIdx.x + threadIdx.x;
+
+    if (i*4 >= k) {
+        return;
+    }
+
+    const int ib = i/(QK4_0/4);
+    const int iqs = i%(QK4_0/4);
+
+    const block_q4_0 * x = (const block_q4_0 *) vx;
+    const uchar2 qs = *(const uchar2 *)(x[ib].qs + iqs*2);
+    const dfloat d = x[ib].d;
+
+    dfloat2 dv0;
+    dv0.x = (int)(qs.x & 0xf) - 8;
+    dv0.y = (int)(qs.y & 0xf) - 8;
```

Comment:
HIP/ROCm treats the `x` and `y` variables of a half2 as shorts, so I think this would work better, and then the same change for dv1 just below this.
```cpp
    #ifdef GGML_CUDA_F16
    dv0 = make_half2((int)(qs.x & 0xf) - 8, (int)(qs.y & 0xf) - 8);
    #else
    dv0.x = (int)(qs.x & 0xf) - 8;
    dv0.y = (int)(qs.y & 0xf) - 8;
    #endif
```

Reply:
Fixed. Introduced a `make_dfloat2` macro to create the proper dfloat2 (half2 or float2)